src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+    }
+
+    public static class Node<T> {
+        private T value;
+        private Node prev;
+        private Node next;
+
+        public Node(T value) {
+            this.value = value;
+        }
+
+        public Node(Node prev, T value, Node next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+        } else {
+            newNode.prev = tail;
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 && index > size) {
+            throw new IndexOutOfBoundsException("No such index exception: index "
+                    + index + " for size " + size);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> oldNode = getNode(index);
+        Node<T> newNode = new Node<>(value);
+        if (oldNode.prev == null) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (oldNode.next == null) {
+            newNode.prev = oldNode.prev;
+            newNode.next = oldNode;
+            oldNode.prev = newNode;
+            oldNode.prev.next = newNode;
+        } else {
+            newNode.prev = oldNode.prev;
+            newNode.next = oldNode;
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexExeption(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexExeption(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexExeption(index);
+        T result = getNode(index).value;
+        unlink(getNode(index));
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (object == null) {
+                if (getNode(i).value == null) {
+                    remove(i);
+                    return true;
+                }
+            } else if (object.equals(getNode(i).value)) {
+                remove(i);
+                return true;
+            }
+        }
```

Comment:
Don't forget about performance. Each 'getNode' method calls loop inside. You can remove node using only 1 loop
```suggestion
        for (Node<T> node = head; node != null; node = node.next) {
            if (node.value == object || object != null && object.equals(node.value)) {
                unlink(node);
                return true;
            }
        }
```